sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -1035,6 +1046,130 @@ private Progress getProgress() {
     return null;
   }
 
+  private HandlesSplits.SplitResult trySplitForWindowObservingTruncateRestriction(
+      double fractionOfRemainder, HandlesSplits splitDelegate) {
```

Comment:
We should be scaling the fraction of the remainder relative to the number of windows we have to find the best split point.

This logic should essentially mirror https://github.com/apache/beam/blob/13c77a8f7a7c726aaa6dadad7812acdf8772ba7c/sdks/python/apache_beam/runners/common.py#L892 except that we will be using the `truncate` transform id, input id, and coder.

I suggest following the approach there where we have a static method that does all the heavy lifting so we can test it well and we have some simple wrappers which pass forward all the necessary arguments to the static method that is visible for testing.

Reply:
That means [trySplitForElementAndRestriction](https://github.com/apache/beam/blob/fa3ca2b11e2ca031232245814389d29c805f79e7/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java#L1039) requires the same change.

Reply:
Yeah, that would make sense and would allow us to share a bunch of code between the truncate split logic and the process sized elements split logic.

Reply:
I see, let's do the scale progress PR first. I'll refactor the code in https://github.com/apache/beam/pull/12430 a little bit.